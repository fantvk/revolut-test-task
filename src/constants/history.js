export const HISTORY_TYPE_REVENUE = 'revenue';
export const HISTORY_TYPE_COST = 'cost';
