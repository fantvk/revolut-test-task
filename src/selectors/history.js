import { createSelector } from '@reduxjs/toolkit';
import { flow, get, filter, orderBy } from 'lodash/fp';

export const getHistory = get(['history']);
export const getEntities = flow(getHistory, get(['entities']));

export const getCurrencyEntities = createSelector(
  getEntities,
  (_, currency) => currency,
  (entities, currency) => flow(
    filter({ currency }),
    orderBy('timestamp', 'desc'),
  )(entities),
);
