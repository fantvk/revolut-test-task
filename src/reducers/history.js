import { createSlice } from '@reduxjs/toolkit';
import { uniqueId } from 'lodash/fp';

import { HISTORY_TYPE_REVENUE } from 'constants/history';

export const slice = createSlice({
  name: 'history',
  initialState: {
    entities: [],
  },
  reducers: {
    createEntity: (state, action) => {
      const { id, timestamp, type, currency, amount, title, balance } = action.payload;
      const entity = {
        id: id || uniqueId(),
        type: type || HISTORY_TYPE_REVENUE,
        timestamp: timestamp || Date.now(),
        currency,
        amount,
        title,
        balance,
      };

      state.entities.push(entity);
    },
  },
});

export const { createEntity } = slice.actions;

export default slice.reducer;