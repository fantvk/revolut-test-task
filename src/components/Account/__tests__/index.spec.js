import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { act, render } from '@testing-library/react';

import { HISTORY_TYPE_COST } from 'constants/history';
import { Account } from '../index';

const mockStore = configureMockStore();
const store = mockStore({
  app: { balance: { EUR: 500000 } },
  history: {
    entities: [{
      id: 1,
      type: HISTORY_TYPE_COST,
      timestamp: 1584183448921,
      currency: 'EUR',
      amount: 10,
      title: 'Title',
      balance: 0,
    }]
  }
});

describe('Account component', () => {
  it('should render ok', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <Account currencies={['EUR', 'USD']} />
        </Provider>
      );

      expect(container).toMatchSnapshot();
    });
  });
});

