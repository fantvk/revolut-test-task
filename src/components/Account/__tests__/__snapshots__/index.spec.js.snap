// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account component should render ok 1`] = `
<div>
  <main
    class="ant-layout-content wrapper"
  >
    <div
      class="breadcrumb ant-breadcrumb"
    >
      <span>
        <span
          class="ant-breadcrumb-link"
        >
          App
        </span>
        <span
          class="ant-breadcrumb-separator"
        >
          /
        </span>
      </span>
      <span>
        <span
          class="ant-breadcrumb-link"
        >
          Account
        </span>
        <span
          class="ant-breadcrumb-separator"
        >
          /
        </span>
      </span>
    </div>
    <div
      class="content"
    >
      <div
        class="ant-row"
      >
        <div
          class="ant-col ant-col-24 col"
        >
          <div
            class="ant-radio-group ant-radio-group-outline"
          >
            <label
              class="ant-radio-button-wrapper ant-radio-button-wrapper-checked"
            >
              <span
                class="ant-radio-button ant-radio-button-checked"
              >
                <input
                  checked=""
                  class="ant-radio-button-input"
                  type="radio"
                  value="EUR"
                />
                <span
                  class="ant-radio-button-inner"
                />
              </span>
              <span>
                EUR
              </span>
            </label>
            <label
              class="ant-radio-button-wrapper"
            >
              <span
                class="ant-radio-button"
              >
                <input
                  class="ant-radio-button-input"
                  type="radio"
                  value="USD"
                />
                <span
                  class="ant-radio-button-inner"
                />
              </span>
              <span>
                USD
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="ant-row"
      >
        <div
          class="ant-col ant-col-24 col"
        >
          <span
            class="ant-tag ant-tag-blue balance"
          >
            €500,000.00
          </span>
        </div>
      </div>
      <div
        class="ant-row"
      >
        <div
          class="ant-col ant-col-12 ant-col-offset-6 col"
        >
          <div
            class="ant-list ant-list-split"
          >
            <div
              class="ant-spin-nested-loading"
            >
              <div
                class="ant-spin-container"
              >
                <ul
                  class="ant-list-items"
                >
                  <li
                    class="ant-list-item entity"
                  >
                    <div
                      class="ant-list-item-meta"
                    >
                      <div
                        class="ant-list-item-meta-avatar"
                      >
                        <span
                          class="ant-avatar ant-avatar-circle ant-avatar-icon"
                        >
                          <span
                            aria-label="sync"
                            class="anticon anticon-sync"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              class=""
                              data-icon="sync"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M168 504.2c1-43.7 10-86.1 26.9-126 17.3-41 42.1-77.7 73.7-109.4S337 212.3 378 195c42.4-17.9 87.4-27 133.9-27s91.5 9.1 133.8 27A341.5 341.5 0 01755 268.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.7 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c0-6.7-7.7-10.5-12.9-6.3l-56.4 44.1C765.8 155.1 646.2 92 511.8 92 282.7 92 96.3 275.6 92 503.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8zm756 7.8h-60c-4.4 0-7.9 3.5-8 7.8-1 43.7-10 86.1-26.9 126-17.3 41-42.1 77.8-73.7 109.4A342.45 342.45 0 01512.1 856a342.24 342.24 0 01-243.2-100.8c-9.9-9.9-19.2-20.4-27.8-31.4l60.2-47a8 8 0 00-3-14.1l-175.7-43c-5-1.2-9.9 2.6-9.9 7.7l-.7 181c0 6.7 7.7 10.5 12.9 6.3l56.4-44.1C258.2 868.9 377.8 932 512.2 932c229.2 0 415.5-183.7 419.8-411.8a8 8 0 00-8-8.2z"
                              />
                            </svg>
                          </span>
                        </span>
                      </div>
                      <div
                        class="ant-list-item-meta-content"
                      >
                        <h4
                          class="ant-list-item-meta-title"
                        >
                          Title
                        </h4>
                        <div
                          class="ant-list-item-meta-description"
                        >
                          3/14/2020, 12:57:28 PM
                        </div>
                      </div>
                    </div>
                    <div>
                      -
                       
                      €10.00
                    </div>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
